### PR TITLE
#408

### PR DESCRIPTION
**Pull request by prix-uniris**
# Description
 This Implements the new VarInt Scheme as mentioned in #408 . This will support for new list sizes with any length while supporting serialization and deserialization for all the structures in blockchain context.

Fixes # (issue)

## Type of change
- New feature (non-breaking change which adds functionality)
- Breaking change (fix or feature that would cause existing functionality to not work as expected)

# How Has This Been Tested?

Test already present in test suite are sufficient to test the code changes

# Checklist:

- My code follows the style guidelines of this project
- I have performed a self-review of my own code
- I have commented my code, particularly in hard-to-understand areas
- I have made corresponding changes to the documentation
- My changes generate no new warnings
- I have added tests that prove my fix is effective or that my feature works
- New and existing unit tests pass locally with my changes
- Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: lib/archethic/utils/varint.ex [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+        message:
+          "the argument value is invalid, Byte Size Supplied: #{bytes}, Bytes found : #{byte_size(rest)}. Should be equal."
+    end
+
+    <<value::size(bytes)-unit(8)>> = rest
+
+    %__MODULE__{
+      bytes: bytes,
+      value: value
+    }
```
Comment (imnik11) [inline lib/archethic/utils/varint.ex:65]: you should add rest in the end, like we do elsewhere for example in end of node sync[here](https://github.com/archethic-foundation/archethic-node/blob/master/lib/archethic/beacon_chain/slot/end_of_node_sync.ex#:~:text=%7D-,%22%22%22,end,-%40spec%20to_map(t)

File: lib/archethic/utils/varint.ex [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+      1..256
+      |> Enum.with_index(fn element, index -> {index + 1, element} end)
+      |> Enum.map(fn {i, x} -> {i, Integer.pow(2, 8 * x)} end)
+
+    # Since Range is in sorted order, first find would be the least amount of bytes required range.
+    {bytes, _range} = ranges |> Enum.find(fn {_bytes, range_max} -> value < range_max end)
+    bytes
+  end
+
+  @spec serialize(__MODULE__.t()) :: <<_::64, _::_*8>>
```
Comment (imnik11) [inline lib/archethic/utils/varint.ex:45]: maybe Improve spec

File: lib/archethic/utils/varint.ex (hunk 1 above)
Comment (prix-uniris) [reply lib/archethic/utils/varint.ex:65]: added a new function get_value which returns the parsed integer and rest.

File: lib/archethic/utils/varint.ex (hunk 2 above)
Comment (prix-uniris) [reply lib/archethic/utils/varint.ex:45]: resolved

File: lib/archethic/p2p/message.ex [hunk 3]
```diff
@@ -208,8 +209,11 @@ defmodule Archethic.P2P.Message do
         welcome_node_public_key: welcome_node_public_key,
         validation_node_public_keys: validation_node_public_keys
       }) do
+    encoded_validation_node_public_keys_length =
```
Comment (Neylix) [inline lib/archethic/p2p/message.ex:212]: As this array cannot overpass a length of 200 we don't need to use VarInt there

File: lib/archethic/p2p/message.ex [hunk 4]
```diff
@@ -239,7 +246,9 @@ defmodule Archethic.P2P.Message do
         },
         confirmed_validation_nodes: confirmed_validation_nodes
       }) do
-    nb_validation_nodes = length(chain_replication_tree)
+    encoded_nb_validation_nodes_length = length(chain_replication_tree) |> VarInt.from_value()
```
Comment (Neylix) [inline lib/archethic/p2p/message.ex:249]: As this array cannot overpass a length of 200 we don't need to use VarInt there

File: lib/archethic/p2p/message.ex (hunk 3 above)
Comment (prix-uniris) [reply lib/archethic/p2p/message.ex:212]: Changed for length(nb_validation_nodes)

File: lib/archethic/p2p/message.ex [hunk 5]
```diff
@@ -221,8 +225,11 @@ defmodule Archethic.P2P.Message do
         io_storage_nodes_view: io_storage_nodes_view,
         previous_storage_nodes_public_keys: previous_storage_nodes_public_keys
       }) do
+    encoded_previous_storage_nodes_public_keys_length =
```
[the remark this answers is not in the document]
Comment (prix-uniris) [reply lib/archethic/p2p/message.ex:228]: Changed for length(nb_previous_storage_nodes)

File: lib/archethic/p2p/message.ex (hunk 4 above)
Comment (prix-uniris) [reply lib/archethic/p2p/message.ex:249]: Changed for length(chain_replication_tree)

File: lib/archethic/transaction_chain/transaction.ex [hunk 6]
```diff
@@ -645,12 +646,14 @@ defmodule Archethic.TransactionChain.Transaction do
       |> Enum.map(&CrossValidationStamp.serialize/1)
       |> :erlang.list_to_binary()
 
+    encoded_cross_validations_stamps_len = length(cross_validation_stamps) |> VarInt.from_value()
```
[the remark this answers is not in the document]
Comment (prix-uniris) [reply lib/archethic/transaction_chain/transaction.ex:649]: Changed in length(cross_validation_stamps)

**Review by Neylix — APPROVED** [review]

**Pull request merged**